### DISCUSSION TITLE
Why is typeshed information not integrated into the official python docs?

### DISCUSSION THREAD

**Discussion by topherbuckley** [python/typing#1227] [Q&A]
I was just introduced to typeshed, and from a brief look, it looks like it is to be used by other tools like PyCharm, mypy, etc, but I'm not seeing anything compiled to a general reference like the online python docs. Is this not a thing; a general reference for types on built-ins? I feel like the info in typeshed should somehow be merged/integrated with the docs. It feels off that there is documentation about type hints not in the official docs.

More info on this [SO Question](https://stackoverflow.com/questions/73163352/python-documentation-input-variable-type-hints)

**Comment by srittau** [comment]
There isn't a definitive answer for this, but there are a few factors:

* typeshed is incomplete, not all stdlib definitions are typed.
* typeshed isn't part of Python's core development. Only one of the active maintainers is a Python Core Developer (and that is a rather recent development).
* typeshed annotations aren't really meant as documentation, they are a technical description meant for consumption by type checkers. Therefore, sometimes type annotation can be hard to understand. The thoughts on this are changing a bit and there have been some thoughts to possibly include more information for human readers, but there is nothing concrete so far.

**Reply by topherbuckley** [reply]
@srittau thanks for following up and the additional info. I'm sure there is a bunch of additional information surrounding this that I'm oblivious to, but just my 10 cents:

>typeshed is incomplete, not all stdlib definitions are typed.

I would argue this shouldn't be a reason to avoid including the parts which are complete. The information that is available is incredibly helpful, and if someone is not using an IDE or the like which makes use of typeshed, they wouldn't even know this information exists (like myself). 

>typeshed annotations aren't really meant as documentation, they are a technical description meant for consumption by type checkers. Therefore, sometimes type annotation can be hard to understand. The thoughts on this are changing a bit and there have been some thoughts to possibly include more information for human readers, but there is nothing concrete so far.

Maybe I just haven't seen the difficult to understand annotation you mention, but I find it incredibly helpful, and I'm sure I'm not alone in that. More details in the SO Question I mentioned, but in short, without this information the docs feel like a recipe asking you to add ingredient A without mentioning exactly what A is.

**Reply by Akuli** [reply]
You can build CPython's docs yourself, just like for any other project that uses sphinx for their docs. Maybe you should make a script that builds docs with type hints, publishes the results somewhere, and see how many people end up using it :)

**Reply by JelleZijlstra** [reply]
> > typeshed is incomplete, not all stdlib definitions are typed.
> 
> I would argue this shouldn't be a reason to avoid including the parts which are complete. The information that is available is incredibly helpful, and if someone is not using an IDE or the like which makes use of typeshed, they wouldn't even know this information exists (like myself).

I also don't think incompleteness is a great argument either; I doubt there's any significant gaps left in our coverage of the stdlib.

> Maybe I just haven't seen the difficult to understand annotation you mention

Some examples are the overloads for map(): https://github.com/python/typeshed/blob/3d9e9078737b8635651d873c39da1350e8d8f738/stdlib/builtins.pyi#L1335

**Reply by topherbuckley** [reply]
@Akuli good idea! I'll add it to the pile of to-dos that may never get done :/ , but definitely something I'd like to try!

@JelleZijlstra although at first glance the `map()` stub is a bit convoluted due to all the short-named generics, and the overloaded `__init__` functions (not sure why they need to explicitly overload 6 times ¯\_(ツ)_/¯, but just reading the first one does already add value to the official documentation. For example, the function and iterable have to be of the same generic type `_S` (i.e. the function can't iterate over `int` and the iterable contain `str`). Borrowing code from [RealPython](https://realpython.com/python-map-function/) as an example:

``` python
def square(number):
    return number ** 2


# numbers = [1, 2, 3, 4, 5]
numbers = ["1", "2", "3", "4", "5"]


squared = map(square, numbers)

print(list(squared))
```

This is a bit of a forced example, but the list of str will yield a static type error in PyCharm as I would expect because of what I read in the typeshed stub (i.e. I knew this would happen before testing it; knowledge I wouldn't otherwise have without the typeshed stub).

**Reply by topherbuckley** [reply]
So I just had a quick look at building the CPython docs, and how one might approach this. Building the CPython docs with Sphinx is relatively straightforward, but how to pull the type hints from typeshed seems a bit more complicated. 

For example, the typeshed doc for [makedirs](https://github.com/python/typeshed/blob/402489b40afffcba11ca9ae8879a8340937829c1/stdlib/os/__init__.pyi#L708) I referenced before lists the type hints as:

``` python
makedirs(name: StrOrBytesPath, mode: int = ..., exist_ok: bool = ...) -> None
```

whereas PyCharm somehow follows `StrOrBytesPath` and displays it as:

``` python
makedirs(name: str | bytes | PathLike[str] | PathLike[bytes],
             mode: int = ...,
             exist_ok: bool = ...) -> None
```

You can find this definition for this TypeAlias in [_typeshed](https://github.com/python/typeshed/blob/402489b40afffcba11ca9ae8879a8340937829c1/stdlib/_typeshed/__init__.pyi#L132) but I'm not sure how to easily do this programmatically for all functions. Is this something you can automate with Sphinx? I'll keep having a look, but guessing there are some sphinx experts around here that might be able to answer this a lot quicker than I can.

**Reply by srittau** [reply]
To clarify: I believe that adding type hints to the Python docs in some way would be a good idea. Using _typeshed as a starting point would use quite a lot of work, though. It would probably still be better exploring this than starting from scratch and needing to maintain a third place (besides the implementation and typeshed).

**Reply by AlexWaygood** [reply]
> guessing there are some sphinx experts around here that might be able to answer this a lot quicker than I can.

@AA-Turner, maybe? :)

**Reply by topherbuckley** [reply]
@srittau yeah, I was a bit surprised to see that the typeshed stubs and python sphinx files all looks to be manually assembled (correct me if I'm wrong) rather than everything being pulled from some central place. I agree that adding a 3rd place of development wouldn't be ideal, so I was trying to find a way to automate the injection of the typeshed type hints into the python documentation either via the sphinx build system or some awk|sed|python text manipulation. Then the final versions (pdf/html/epub/etc) from sphinx could be built with github actions or the like and simply be made from existing sources rather than maintaining a separate one. 

@AlexWaygood you see any easy way to do what I mention above via sphinx (text injection) or will this require some awk magic?

**Reply by Gobot1234** [reply]
To inject the type hints into python code you might be able to use https://github.com/google/pytype/tree/main/pytype/tools/merge_pyi

**Reply by AlexWaygood** [reply]
> @AlexWaygood you see any easy way to do what I mention above via sphinx (text injection) or will this require some awk magic?

No point tagging me, I don't know anything about sphinx :) I just tagged a friendly sphinx expert who might be able to help, if he has the time!

**Reply by topherbuckley** [reply]
@Gobot1234 thanks for the hint. I'll check it out. 

@AlexWaygood, sorry just a slip of the autocomplete. I intended to tag @AA-Turner. Understood on that. Happy to receive help, but happy to tinker away until then when time is available.

**Reply by adam-grant-hendry** [reply]
This is perhaps also related to [my question](https://github.com/python/typing/discussions/1255).
